### PR TITLE
Feat!: add support for heredoc strings (Postgres, ClickHouse)

### LINKED ISSUE #2316: clickhouse: heredoc is not parsable
```python
import sqlglot

sql = """
SELECT $$this is a heredoc string$$
"""
print(sqlglot.parse_one(sql, read="clickhouse"))
```
```
sqlglot.errors.ParseError: Invalid expression / Unexpected token. Line 2, Col: 18.
  
SELECT $$this is a heredoc string$$
```

docs: https://clickhouse.com/docs/en/sql-reference/syntax#heredoc

_Note: `$$anything$$` is a form of heredoc too, i.e. `SELECT $$aaa$$this is a heredoc string$$aaa$$`, and it could be multiline, etc._

### PR DESCRIPTION
**Pull request by GeorgeSittas**
Fixes #2316 - planning to do another pass on this and double-check some changed tests etc.

cc: @pkit

References:
- https://clickhouse.com/docs/en/sql-reference/syntax#heredoc
- https://www.postgresql.org/docs/8.1/sql-syntax.html (see «Dollar-Quoted String Constants» section)

### CODE REVIEW

File: sqlglot/dialects/postgres.py [hunk 1]
```diff
@@ -248,11 +248,10 @@ class Postgres(Dialect):
     }
 
     class Tokenizer(tokens.Tokenizer):
-        QUOTES = ["'", "$$"]
```
Comment (tobymao) [inline sqlglot/dialects/postgres.py:251]: this was incorrect before?

Comment (GeorgeSittas) [reply sqlglot/dialects/postgres.py:251]: more like partially correct; it didn't take into account tagged heredocs like the following:

```
georgesittas=# select $foo$this is a string$foo$;
     ?column?
------------------
 this is a string
(1 row)

Comment (GeorgeSittas) [reply sqlglot/dialects/postgres.py:251]: FYI the relevant commit is https://github.com/tobymao/sqlglot/pull/617/files, cc @mpf82

File: tests/dialects/test_postgres.py [hunk 2]
```diff
@@ -207,7 +207,6 @@ def test_postgres(self):
         self.validate_identity("SELECT ARRAY[1, 2, 3] @> ARRAY[1, 2]")
         self.validate_identity("SELECT ARRAY[1, 2, 3] <@ ARRAY[1, 2]")
         self.validate_identity("SELECT ARRAY[1, 2, 3] && ARRAY[1, 2]")
-        self.validate_identity("$x")
```
Comment (GeorgeSittas) [inline tests/dialects/test_postgres.py:210]: This is invalid postgres:

```
georgesittas=# select $x;
ERROR:  syntax error at or near "$"
LINE 1: select $x;
```

File: tests/dialects/test_redshift.py [hunk 3]
```diff
@@ -226,7 +226,6 @@ def test_identity(self):
         self.validate_identity("SELECT * FROM #x")
         self.validate_identity("SELECT INTERVAL '5 day'")
         self.validate_identity("foo$")
-        self.validate_identity("$foo")
```
Comment (GeorgeSittas) [inline tests/dialects/test_redshift.py:229]: This is invalid Redshift
Screenshot 2023-09-26 at 8 48 57 PM

**Review by tobymao — APPROVED** [review]

**Comment by pkit** [comment]
Yup, works fine for my use case. Thanks!

**Pull request merged**